lib/archethic/mining/pending_transaction_validation.ex

```diff
@@ -825,49 +832,129 @@ defmodule Archethic.Mining.PendingTransactionValidation do
+    results =
+      [tx_genesis_result, ref_genesis_result, ref_tx_result] =
+      Task.yield_many(tasks)
+      |> Enum.map(fn {task, res} ->
+        res || Task.shutdown(task, :brutal_kill)
+      end)
+
+    with 0 <- Enum.count(results, &(&1 == nil)),
+         {:ok, {:ok, genesis_address}} <- tx_genesis_result,
+         {:ok, {:ok, ^genesis_address}} <- ref_genesis_result,
+         {:ok, {:ok, %Transaction{data: %TransactionData{content: content}}}} <- ref_tx_result,
          {:ok, reference_json_token} <- Jason.decode(content),
          %{"type" => "fungible", "allow_mint" => true} <- reference_json_token do
       :ok
     else
+      i when is_integer(i) ->
+        {:error, "Timeout when fetching the reference token or the genesis address"}
+
+      {:exit, _} ->
+        {:error, "Error when fetching the reference token or the genesis address"}
+
       %{"type" => "non-fungible"} ->
         {:error, "Invalid token transaction - token_reference must be fungible"}
 
       %{"type" => "fungible"} ->
         {:error, "Invalid token transaction - token_reference does not have allow_mint: true"}
 
-      {:ok, _} ->
+      {:ok, {:ok, _}} ->
         {:error,
          "Invalid token transaction - token_reference is not in the same transaction chain"}
 
-      {:error, :transaction_not_exists} ->
+      {:ok, {:error, :transaction_not_exists}} ->
         {:error, "Invalid token transaction - token_reference not found"}
 
-      {:error, :invalid_transaction} ->
+      {:ok, {:error, :invalid_transaction}} ->
         {:error, "Invalid token transaction - token_reference is invalid"}
 
-      {:error, :network_issue} ->
+      {:ok, {:error, :network_issue}} ->
         {:error, "A network issue was raised, please retry later"}
 
       {:error, %Jason.DecodeError{}} ->
         {:error,
          "Invalid token transaction - token_reference exists but does not contain a valid JSON"}
+    end
+  end
 
-      :error ->
-        {:error, "Invalid token transaction - token_reference is not an hexadecimal"}
+  defp verify_token_recipients(json_token) do
+    case json_token["recipients"] || [] do
+      [] ->
+        :ok
+
+      recipients ->
+        with :ok <- validate_token_recipients_amounts(recipients, json_token),
+             :ok <- validate_token_recipients_total(recipients, json_token) do
+          validate_token_recipients_addresses(recipients)
+        end
+    end
+  end
+
+  defp validate_token_recipients_addresses(recipients) do
+    valid? =
+      recipients
+      |> Enum.map(& &1["to"])
+      |> Enum.map(&Base.decode16!(&1, case: :mixed))
+      |> Enum.all?(&Crypto.valid_address?/1)
+
+    if valid? do
+      :ok
+    else
+      {:error, "Invalid token transaction - invalid recipients addresses"}
+    end
+  end
+
+  defp validate_token_recipients_total(recipients, json_token) do
+    total =
+      recipients
+      |> Enum.map(& &1["amount"])
+      |> Enum.sum()
+
+    if total <= json_token["supply"] do
```

Comment:
You can retrieve suply by pattern matching function paramter to stay consistent with entire project